core/services/relay/evm/chain_reader.go

```diff
@@ -0,0 +1,80 @@
+package evm
+
+import (
+	"context"
+	"fmt"
+
+	relaytypes "github.com/smartcontractkit/chainlink-relay/pkg/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/types"
+)
+
+// constructor for ChainReader, returns nil if there is any error
+func newChainReader(lggr logger.Logger, chain evm.Chain, ropts *types.RelayOpts) *chainReader {
```

Comment:
I am finding this difficult to read, and it would be hard to test as well. Why don't we use a pure constructor here:
```suggestion
func newChainReader(lggr, logger.Logger, chain evm.Chain, ropts *types.RelayOpts) (*chainReader, error) {
```
The callers can decide how to handle the error.